### PR TITLE
[SPARK-17114][SQL] Fix aggregates grouped by literals with empty input.

### PR DESCRIPTION
**Pull request by hvanhovell**
## What changes were proposed in this pull request?

This PR fixes an issue with aggregates that have an empty input, and use a literals as their grouping keys. These aggregates are currently interpreted as aggregates **without** grouping keys, this triggers the ungrouped code path (which aways returns a single row).

This PR fixes this by adding a `isGrouped` flag to `Aggregate`. This flag is inferred when the `Aggregate` is created, but is kept when we change the `Aggregate`s grouping expressions. I have tried to implement this by wrapping the `groupingExpressions` in an `Option` but this is much more invasive and leads to much harder to use code.
## How was this patch tested?

Added tests to `SQLQueryTestSuite`.

### CODE REVIEW

**Comment by hvanhovell** [comment]
cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 1]
```diff
@@ -484,7 +494,10 @@ case class Aggregate(
       }.nonEmpty
     )
 
-    !expressions.exists(!_.resolved) && childrenResolved && !hasWindowExpressions
+    !expressions.exists(!_.resolved) &&
+      childrenResolved &&
+      !hasWindowExpressions &&
+      (isGrouped || groupingExpressions.isEmpty)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: why have this condition?

**Comment by davies** [comment]
@hvanhovell What's the reason that we can not just use the literal use grouping key?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 1 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: It does not matter if `isGrouped` has any grouping expressions (literal grouping expressions are eliminated during optimization). It is however problematic when a not-grouped `Aggregate` has grouping expressions; this means that we have not derived the `isGrouped` flag  correctly.

**Comment by hvanhovell** [comment]
@davies literal grouping keys get eliminated during optimization. We could modify the optimization rule, but I do feel this is more robust and it is more clear on what we intent to do. It probably also saves a few cycles during query execution.

**Comment by davies** [comment]
@hvanhovell  Since the optimization rule change the result (or sematics), I'd like to fix it, not to introduce more complexcity in physical layer. The saved few cycles may not worth, because people usually not use that in practice.

**Comment by hvanhovell** [comment]
closing this in favor of https://github.com/apache/spark/pull/15101.

**Pull request closed without merging**